pydantic/main.py

```diff
@@ -270,6 +280,9 @@ def dict(
         get_key = partial(get_key, self.fields)
 
         allowed_keys = self._calculate_keys(include=include, exclude=exclude, skip_defaults=skip_defaults)
+        if as_type is not None:
+            model_keys = set(as_type.__fields__)
+            allowed_keys = model_keys if allowed_keys is None else allowed_keys.intersection(model_keys)
```

Comment:
As far as I can see `as_type` just restricts the fields dumped to match the fields on the `as_type` model. Is that right?

If that's the case:
1. couldn't it be done outside `.dict()`, and just use `.dict(include=set(AsTypeModel.__fields__))`?
2. I don't see that this gives any guarantee about field types, what happens if I override a field on a child model with a new type?

Reply:
@samuelcolvin The problem is with deeply nested models, how would I specify the include types?

Think about this example:
```python
import datetime
from typing import List, Dict, NewType, Tuple
import uuid

from pydantic import BaseModel, dump_as_type

UserID = NewType("UserID", uuid.UUID)
RegistrationTime = NewType("RegistrationTime", datetime.datetime)


class UserAPI(BaseModel):
    username: str
    email: str


class UserInDatabase(UserAPI):
    hashed_password: str


class AppData(BaseModel):
    items: List[str]
    users: Dict[UserID, Tuple[UserAPI, RegistrationTime]]
```
Is there an easy way to serialize `AppData` if `users` contains `UserInDatabase` in the first element of the tuples? (Note: tuple is covariant, so there isn't even a typesafety violation if UserInDatabase instances were present)

Reply:
I believe it currently handles overrides properly -- it uses the type as specified for the field in the `as_type` argument to `dump_as_type`. This is tested in the current test, though I think deserves a little more comprehensive testing to make sure it always behaves properly.

Reply:
really, I don't see any call to `dump_as_type` except in tests?

Reply:
> As far as I can see as_type just restricts the fields dumped to match the fields on the as_type model. Is that right?

It also goes one step farther -- it does this in a *nested* way, which as far as I can tell is not possible to do with the current APIs.

Reply:
makes sense, resolving this discussion to avoid duplication with the issue.